thirdparty/hnswlib/hnswlib/hnswalg.h

```diff
@@ -226,22 +289,52 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
 
     inline dist_t
     calcDistance(const tableint id1, const tableint id2) const {
-        dist_t dist = fstdistfunc_(getDataByInternalId(id1), getDataByInternalId(id2), dist_func_param_);
-        if (metric_type_ == Metric::COSINE) {
-            dist /= (data_norm_l2_[id1] * data_norm_l2_[id2]);
+        if (sq_enabled_) {
+            return fstdistfunc_sq_(getSQDataByInternalId(id1), getSQDataByInternalId(id2), dist_func_param_) * alpha_ *
+                   alpha_ / 127.0f / 127.0f;
+        } else {
+            dist_t dist = fstdistfunc_(getDataByInternalId(id1), getDataByInternalId(id2), dist_func_param_);
+            if (metric_type_ == Metric::COSINE) {
+                dist /= (data_norm_l2_[id1] * data_norm_l2_[id2]);
+            }
+            return dist;
         }
-        return dist;
     }
 
     inline dist_t
     calcDistance(const void* vec, const tableint id) const {
+        if (sq_enabled_) {
+            return fstdistfunc_sq_(vec, getSQDataByInternalId(id), dist_func_param_) * alpha_ * alpha_ / 127.0f /
+                   127.0f;
+        } else {
+            dist_t dist = fstdistfunc_(vec, getDataByInternalId(id), dist_func_param_);
+            if (metric_type_ == Metric::COSINE) {
+                dist /= data_norm_l2_[id];
+            }
+            return dist;
+        }
+    }
+
+    inline dist_t
+    calcRefineDistance(const void* vec, const tableint id) const {
         dist_t dist = fstdistfunc_(vec, getDataByInternalId(id), dist_func_param_);
         if (metric_type_ == Metric::COSINE) {
             dist /= data_norm_l2_[id];
         }
         return dist;
     }
 
+    void
+    prefetchData(const tableint id) const {
+#if defined(USE_PREFETCH)
+        if (sq_enabled_) {
+            _mm_prefetch(getSQDataByInternalId(id), _MM_HINT_T0);
+        } else {
+            _mm_prefetch(getDataByInternalId(id), _MM_HINT_T0);
+        }
```

Comment:
just mentioning that this is x86-only code, unless `USE_PREFETCH` is enabled for x86 only as well

Reply:
We need to centralized these platform related code just like what we do for distance hook. Created an issue to track https://github.com/zilliztech/knowhere/issues/416.

Reply:
agree

Reply:
using the hook mechanism for prefetching is a horrible idea